blockchain/client.go

```diff
@@ -25,11 +31,73 @@ func NewClient(url string) (*Client, error) {
 		return nil, err
 	}
 
+	subs := make(map[string]chan *pallets.Events)
+	subs["DdcClusters"] = make(chan *pallets.Events)
+	subs["DdcCustomers"] = make(chan *pallets.Events)
+	subs["DdcPayouts"] = make(chan *pallets.Events)
+
 	return &Client{
 		SubstrateAPI: substrateApi,
-		DdcClusters:  pallets.NewDdcClustersApi(substrateApi),
-		DdcCustomers: pallets.NewDdcCustomersApi(substrateApi, meta),
+		subs:         subs,
+		DdcClusters: pallets.NewDdcClustersApi(
+			substrateApi,
+			subs["DdcClusters"],
+		),
+		DdcCustomers: pallets.NewDdcCustomersApi(substrateApi, meta, subs["DdcCustomers"]),
 		DdcNodes:     pallets.NewDdcNodesApi(substrateApi, meta),
-		DdcPayouts:   pallets.NewDdcPayoutsApi(substrateApi, meta),
+		DdcPayouts:   pallets.NewDdcPayoutsApi(substrateApi, meta, subs["DdcPayouts"]),
 	}, nil
 }
+
+func (c *Client) StartEventsListening() (func(), <-chan error, error) {
+	meta, err := c.RPC.State.GetMetadataLatest()
+	if err != nil {
+		return nil, nil, err
+	}
+	key, err := types.CreateStorageKey(meta, "System", "Events", nil)
+	if err != nil {
+		return nil, nil, err
+	}
+	sub, err := c.RPC.State.SubscribeStorageRaw([]types.StorageKey{key})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	done := make(chan struct{})
+	errCh := make(chan error)
+
+	go func() {
+		for {
+			select {
+			case <-done:
+				return
+			case set := <-sub.Chan():
+				for _, change := range set.Changes {
+					if !codec.Eq(change.StorageKey, key) || !change.HasStorageData {
+						continue
+					}
+
+					events := &pallets.Events{}
+					err = types.EventRecordsRaw(change.StorageData).DecodeEventRecords(meta, events)
+					if err != nil {
+						errCh <- fmt.Errorf("events listener: %w", err)
+					}
+
+					for _, module := range c.subs {
```

Comment:
hm. do i correctly understand that we are replicating events to all subs here?

Reply:
All block events are stored in [one large structure](https://github.com/Cerebellum-Network/cere-ddc-sdk-go/pull/77/files#diff-ad49ad6045bda9c5c2ea7414fa02ef1a9c48164184b9007c567c280f344a9bb0R9) (substrate API library recommended method) and here we pass it by pointer to each pallet. For a smaller structure best practices tell us to copy it while read only access is expected, but this one contains all events of the block and it can be too large to copy. 

Also we can have no events filtering by pallet here because pallets will only pick events then need.